// @suprsend/node-sdk/dist/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SuprsendError = void 0;
exports.base64Encode = base64Encode;
exports.epoch_milliseconds = epoch_milliseconds;
exports.has_special_char = void 0;
exports.is_empty = is_empty;
exports.is_object = is_object;
exports.is_string = is_string;
exports.resolveTilde = resolveTilde;
exports.uuid = uuid;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

var _os = _interopRequireDefault(require("os"));

var _fs = _interopRequireDefault(require("fs"));

var _uuid = require("uuid");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function base64Encode(file) {
  var body = _fs["default"].readFileSync(file);

  return body.toString("base64");
}

function resolveTilde(filePath) {
  if (!filePath || typeof filePath !== "string") {
    return "";
  }

  if (filePath.startsWith("~/") || filePath === "~") {
    return filePath.replace("~", _os["default"].homedir());
  }

  return filePath;
}

var SuprsendError = /*#__PURE__*/function (_Error) {
  (0, _inherits2["default"])(SuprsendError, _Error);

  var _super = _createSuper(SuprsendError);

  function SuprsendError(message) {
    var _this;

    (0, _classCallCheck2["default"])(this, SuprsendError);
    _this = _super.call(this, message);
    _this.name = "SuprsendError";
    return _this;
  }

  return SuprsendError;
}( /*#__PURE__*/(0, _wrapNativeSuper2["default"])(Error));

exports.SuprsendError = SuprsendError;

function is_string(value) {
  return typeof value === "string";
}

function is_object(value) {
  return (0, _typeof2["default"])(value) === "object" && !Array.isArray(value) && value !== null;
}

function is_empty(value) {
  if (is_object(value)) {
    return Object.keys(value) <= 0;
  } else if (Array.isArray(value)) {
    return value.length <= 0;
  }
}

var has_special_char = function has_special_char(str) {
  var _str$toLowerCase;

  return str.startsWith("$") || (str === null || str === void 0 ? void 0 : (_str$toLowerCase = str.toLowerCase()) === null || _str$toLowerCase === void 0 ? void 0 : _str$toLowerCase.startsWith("ss_"));
};

exports.has_special_char = has_special_char;

function uuid() {
  return (0, _uuid.v4)();
}

function epoch_milliseconds() {
  return Math.round(Date.now());
}
